hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/connector/map/MetadataResolver.java

```diff
@@ -56,59 +60,113 @@ public Mapping resolve(String iMapName) {
             return null;
         }
 
-        boolean hd = container.getMapConfig().getInMemoryFormat() == InMemoryFormat.NATIVE;
-        Metadata metadata = hd ? resolveFromHd(container) : resolveFromHeap(iMapName, context);
+        boolean hd = container.getMapConfig().getInMemoryFormat() == InMemoryFormat.NATIVE
+                && !container.getMapConfig().getTieredStoreConfig().isEnabled();
+
+        Metadata metadata = hd
+                ? resolveFromContentsHd(iMapName, context)
+                : resolveFromContents(iMapName, context);
         return metadata == null
                 ? null
                 : new Mapping(iMapName, iMapName, null, TYPE_NAME, null, metadata.fields(), metadata.options());
     }
 
     @Nullable
-    @SuppressWarnings("rawtypes")
-    private Metadata resolveFromHd(MapContainer container) {
-        if (container.getIndexes() == null) {
-            return null;
+    private Metadata resolveFromContents(String name, MapServiceContext context) {
+        for (PartitionContainer partitionContainer : context.getPartitionContainers()) {
+            RecordStore<?> recordStore = partitionContainer.getExistingRecordStore(name);
```

Comment:
what if this nodes record-stores are empty? Can this be an issue?

Reply:
No. Management Center will ask all nodes until it finds one with data. Looks for me like some premature optimization.

Reply:
Or at least we tried the operation to be as lightweight as possible. There is also no higher-level API to access serialised data (`MutationObserver` is also on a low level)